core/primitives/src/stateless_validation.rs

```diff
@@ -24,16 +25,31 @@ type SignatureDifferentiator = String;
 #[derive(Debug, Clone, PartialEq, Eq, BorshSerialize, BorshDeserialize)]
 pub struct EncodedChunkStateWitness(Box<[u8]>);
 
+pub type ChunkStateWitnessSize = usize;
+
 impl EncodedChunkStateWitness {
-    pub fn encode(witness: &ChunkStateWitness) -> Self {
-        Self(borsh::to_vec(witness).expect("borsh serialization should not fail").into())
+    /// Borsh-serialize and compress state witness.
+    /// Returns encoded witness along with the raw (uncompressed) witness size.
+    pub fn encode(witness: &ChunkStateWitness) -> std::io::Result<(Self, ChunkStateWitnessSize)> {
+        const STATE_WITNESS_COMPRESSION_LEVEL: i32 = 3;
+        let borsh_bytes = borsh::to_vec(witness)?;
+        Ok((
+            Self(zstd::encode_all(borsh_bytes.as_slice(), STATE_WITNESS_COMPRESSION_LEVEL)?.into()),
+            borsh_bytes.len(),
+        ))
     }
 
-    pub fn decode(&self) -> Result<ChunkStateWitness, std::io::Error> {
-        ChunkStateWitness::try_from_slice(&self.0)
+    /// Decompress and borsh-deserialize encoded witness bytes.
+    /// Returns decoded witness along with the raw (uncompressed) witness size.
+    pub fn decode(&self) -> std::io::Result<(ChunkStateWitness, ChunkStateWitnessSize)> {
+        // We want to limit the size of decompressed data to address "Zip bomb" attack.
+        const MAX_WITNESS_SIZE: usize = 64_000_000;
+        let borsh_bytes = decompress_with_limit(self.0.as_ref(), MAX_WITNESS_SIZE)?;
+        let witness = ChunkStateWitness::try_from_slice(&borsh_bytes)?;
```

Comment:
Similarly, I would opt for a reader interface here, to avoid having 3 full copies of the data (although one of them is compressed, but still) in memory at the same time. You can still limit the amount of data decompressed by having a layer of `Read` implementation that counts the bytes and passes the data along to the outer reader up to the limit.

Reply:
Same as [this](https://github.com/near/nearcore/pull/10715#discussion_r1563132474).